src/watchmaker/status/__init__.py

```diff
@@ -0,0 +1,63 @@
+# -*- coding: utf-8 -*-
+"""Status Module."""
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals, with_statement)
+
+import logging
+
+import watchmaker.config.status as status_config
+from watchmaker.status.providers.aws import AWSStatusProvider
+from watchmaker.status.providers.azure import AzureStatusProvider
+from watchmaker.utils.imds.detect import provider
+
+
+class Status():
+    """Status factory for providers."""
+
+    def __init__(self, config=None, excluded=None, logger=None):
+        self.logger = logger or logging.getLogger(__name__)
+        self.status_provider = None
+        self.identifier = None
+        self.targets = None
+        self.initialize(config, excluded)
+
+    def initialize(self, config=None, excluded=None):
+        """Initialize and get provider."""
+        if not self.identifier or not self.status_provider:
+            self.identifier = provider(excluded)
+            self.status_provider = \
+                self.__get_status_provider_by_id()
+
+        if not self.targets and config:
+            self.targets = \
+                status_config.get_targets_by_target_type(
+                    config, self.identifier)
+
+    def get_status_provider_identifier(self):
+        """Get provider identifier."""
+        return self.identifier
+
+    def get_status_provider(self):
+        """Get provider."""
+        return self.status_provider
+
+    def tag_resource(self, status):
+        """Tag resource with key and status provided."""
+        if not self.status_provider or not self.targets:
```

Comment:
Why would we call this function such that we would need to gate this?

Reply:
TL/DR;  you might not be on a valid provider, or for this provider you might not have a status target.  

If we returned the provider to watchmaker, every time we went to use the provider we would have to check if the provider was set, think of the Status class as a Proxy for tagging resources, it maintains what the provider and targets are for the platform/instance watchmaker is run on.  This instance does not have to be AWS or AZURE, so the provider would be None, or it may be AWS but there is no status config for this provider so the targets are none.

Reply:
Does that mean we are calling this function even when the config would have nothing to actually tag?

Reply:
Yes.  We always call tag resource, but we don't always call provider.tag_resource. The proxy alleviate needing  conditionals wrapping the call or some declarative dict, which is essentially a conditional of a different means which the proxy is as well.

Could we check if config_status exists or not before calling, yes, that would be conditional above, but then we would still need to do the check inside of tag resource because it might not have targets for this provider, but it does have status configs.  Thats why I did all of the conditions in the call instead of every place I make the call